Sources/Core/Shared/GroupOperation.swift

```diff
@@ -339,3 +448,117 @@ public struct WillAddChildObserver: GroupOperationWillAddChildObserver {
         didAttachToOperation?(operation: operation)
     }
 }
+
+private extension GroupOperation {
+    /**
+     The group operation handles thread-safe addition of operations by utilizing two final operations:
+     - a CanFinishOperation which manages handling GroupOperation internal state and has every child
+       operation as a dependency
+     - a finishingOperation, which has the CanFinishOperation as a dependency
+
+     The purpose of this is to handle the possibility that GroupOperation.addOperation() or
+     GroupOperation.queue.addOperation() are called right after all current child operations have
+     completed (i.e. after the CanFinishOperation has been set to ready), but *prior* to being able
+     to process that the GroupOperation is finishing (i.e. prior to the CanFinishOperation executing and
+     acquiring the GroupOperation.groupFinishLock to set state).
+     */
+    private class CanFinishOperation: NSOperation {
+        private weak var parent: GroupOperation?
```

Comment:
Just thinking, would this not be better as `unowned` rather than weak. This would avoid having to unwrap it all the time.